lib/WUI/espif.c

```diff
@@ -0,0 +1,521 @@
+                log_error(ESPIF, "MAC len %d too much, reducing to %d", mac_len, MAC_DATA_MAX_LEN);
+                mac_len = MAC_DATA_MAX_LEN;
+            }
+            mac_read = 0;
+            state = MACData;
+            log_debug(ESPIF, "Reading MAC len: %d", mac_len);
+            // No break, with zero len MAC data there might be no other input byte
+            // to trigger MAC processing.
+
+        case MACData:
+            while (c < end && mac_read < mac_len) {
+                // log_debug(ESPIF, "Read MAC byte at %d: %02x", mac_read, *c);
+                mac_data[mac_read++] = *c++;
+            }
+            if (mac_read == mac_len) {
+                process_mac(mac_len, mac_data, netif);
+                state = Intron;
+            }
+            break;
+
+        case Packet:
+            rx_len_read = 0;
+            state = PacketLen;
+            break;
+
+        case PacketLen:
+            if (rx_len_read < sizeof(rx_len)) {
+                ((uint8_t *)&rx_len)[rx_len_read++] = *c++;
+            } else {
+                log_debug(ESPIF, "Reading packet size: %d", rx_len);
+#if ETH_PAD_SIZE
+                rx_len += ETH_PAD_SIZE; /* allow room for Ethernet padding */
+#endif
+                rx_buff = pbuf_alloc(PBUF_RAW, rx_len, PBUF_POOL);
+                if (rx_buff) {
+                    rx_buff_cur = rx_buff;
+                    rx_read = 0;
+                    state = PacketData;
+                } else {
+                    log_error(ESPIF, "Dropping packet due to out of RAM");
+                    state = Intron;
+                }
+            }
+            break;
+
+        case PacketData: {
+            // Copy input to current pbuf (until end of input or current pbuf)
+            const uint32_t to_read = min(rx_buff_cur->len - rx_read, end - c);
+            memcpy(rx_buff_cur->payload + rx_read, c, to_read);
+            c += to_read;
+            rx_read += to_read;
+
+            // Switch to next pbuf
+            if (rx_read == rx_buff_cur->len) {
+                rx_buff_cur = rx_buff_cur->next;
+                rx_read = 0;
+            }
+
+            // Filled all pbufs in a packet (current set to next = NULL)
+            if (!rx_buff_cur) {
+                log_debug(ESPIF, "Read packet size: %d", rx_len);
+                if (netif->input(rx_buff, netif) != ERR_OK) {
+                    log_error(ESPIF, "ethernetif_input: IP input error");
+                    pbuf_free(rx_buff);
+                    state = Intron;
+                    break;
+                }
+                log_debug(ESPIF, "Input packet processed ok");
+                state = Intron;
+            }
+        } break;
+        default:
+            log_error(ESPIF, "Unknow input parser state: %d", state);
+            state = Intron;
+        }
+    }
+    log_debug(ESPIF, "Processed %d from UART", size);
+}
+
+static err_t low_level_output(struct netif *netif, struct pbuf *p) {
```

Comment:
This is supposed to be called from LwIP internal thread. It is guaranteed to be just one. Anyway a mutex was added to prevent low_level_output and ap_join command from mixing with each other that also allows multithreaded output.